nextflow.config

```diff
@@ -1,112 +1,132 @@
 
+
+// Manifest
 manifest {
     name            = 'nf-encyclopedia'
     author          = 'William E Fondrie'
-    homePage        = 'https://github.com/nf-core/encyclopedia'
+    homePage        = 'https://github.com/nf-encyclopedia'
     description     = 'Analyze DIA proteomics data with EncyclopeDIA'
     mainScript      = 'main.nf'
     nextflowVersion = '!>=21.10.3'
 }
+
+
+// Export these variables to prevent local Python/R libraries from conflicting
+// with those in the container The JULIA depot path has been adjusted to a
+// fixed path `/usr/local/share/julia` that needs to be used for packages in
+// the container. See https://apeltzer.github.io/post/03-julia-lang-nextflow/
+// for details on that. Once we have a common agreement on where to keep Julia
+// packages, this is adjustable.
+// Copied from the nf-core template.
+env {
+    PYTHONNOUSERSITE = 1
+    R_PROFILE_USER   = "/.Rprofile"
+    R_ENVIRON_USER   = "/.Renviron"
+    JULIA_DEPOT_PATH = "/usr/local/share/julia"
+}
+
+// Capture exit codes from upstream processes when piping
+process.shell = ['/bin/bash', '-euo', 'pipefail']
+def trace_timestamp = new java.util.Date().format( 'yyyy-MM-dd_HH-mm-ss')
+timeline {
+    enabled = true
+    file    = "${params.report_dir}/execution_timeline_${trace_timestamp}.html"
+}
+report {
+    enabled = true
+    file    = "${params.report_dir}/execution_report_${trace_timestamp}.html"
+}
+trace {
+    enabled = true
+    file    = "${params.report_dir}/execution_trace_${trace_timestamp}.txt"
+}
+dag {
+    enabled = true
+    file    = "${params.report_dir}/pipeline_dag_${trace_timestamp}.html"
+}
+
+// Load base.config by default for all pipelines
+includeConfig 'conf/base.config'
+
+// Function to ensure that resource requirements don't go beyond
+// a maximum limit. Copied from the nf-core template.
+def check_max(obj, type) {
+    if (type == 'memory') {
+        try {
+            if (obj.compareTo(params.max_memory as nextflow.util.MemoryUnit) == 1)
+                return params.max_memory as nextflow.util.MemoryUnit
+            else
+                return obj
+        } catch (all) {
+            println "   ### ERROR ###   Max memory '${params.max_memory}' is not valid! Using default value: $obj"
+            return obj
+        }
+    } else if (type == 'time') {
+        try {
+            if (obj.compareTo(params.max_time as nextflow.util.Duration) == 1)
+                return params.max_time as nextflow.util.Duration
+            else
+                return obj
+        } catch (all) {
+            println "   ### ERROR ###   Max time '${params.max_time}' is not valid! Using default value: $obj"
+            return obj
+        }
+    } else if (type == 'cpus') {
+        try {
+            return Math.min( obj, params.max_cpus as int )
+        } catch (all) {
+            println "   ### ERROR ###   Max cpus '${params.max_cpus}' is not valid! Using default value: $obj"
+            return obj
+        }
+    }
+}
```

Comment:
It seems a little weird to have these functions in the configuration file. Is there another place these could go? Maybe in launchpad when people are launching jobs?

Reply:
Oh I totally agree, but I copied this directly from the nf-core template. Alas, I guess this is what folks are used to seeing in nf-core pipelines (ex: https://github.com/nf-core/sarek/blob/master/nextflow.config)

I don't really understand why they need to define the function here, rather than somewhere else though...

Reply:
This maybe something that we can change in the future.

Reply:
Yes - I think I'm going to leave it like this for now, but you're right: It's ugly.